### PR TITLE
Fixed edge event loss for kafka edges during sync

### PR DESCRIPTION
**Pull request by AndriiLandiak**
## Pull Request description

On Kafka-based edges, an edge event can be silently dropped if it gets polled from the edge-event topic right as an edge sync (or high-priority processing) starts. The event is read from Kafka, skipped without being committed, and the consumer position moves past it — so it's never re-delivered to that session. This PR adds a small "only poll when the session is ready" gate so the event stays in the queue until we can actually process it.

### How it shows up

`DashboardClientTest.testSendDashboardToCloud` has been flaky (~50% on the Kafka edge profile). The test unassigns the edge from a customer and deletes that customer on the cloud, then waits for the customer to disappear on the edge. Every so often the `CUSTOMER/DELETED` event never reaches the edge, the customer is never removed there, and the `getCustomerById(...).isEmpty()` wait times out after 30s.

The in-memory/Postgres edge never reproduces it: `PostgresEdgeGrpcSession` re-reads events by `seqId` from the DB on every iteration, so an interrupted read is simply re-fetched next time around. The Kafka consumer, on the other hand, is created once and polls forward continuously — so a skipped batch is just gone.

### Root cause

`KafkaEdgeGrpcSession.processMsgs(...)` polls a batch and then returns without committing whenever the session isn't ready:

```java
if (!isConnected() || isSyncInProgress() || isHighPriorityProcessing) {
    log.debug("... Skipping iteration");
    return;                 // batch already polled; position advanced; nothing committed
}
...
if (!isInterrupted) {
    consumer.commit();
}
```

The catch: the consumer has *already* polled, so its in-memory position is past that batch. Not committing only makes the *committed* offset lag — the next `poll()` still reads forward, so within that session the skipped events are never seen again.

Here's the exact sequence from a failing run (cloud side, edge `edge-kafka`):

| Time | Event |
|------|-------|
| 00:49:55,035 | `CUSTOMER/DELETED` pushed to the edge-event topic |
| 00:49:55,036 | poll #1 → processes `UNASSIGNED_FROM_CUSTOMER` (a config change) |
| 00:49:55,044 | the config change triggers a full re-sync → `syncInProgress = true`, "General processing interrupted!" |
| 00:49:55,050 | poll #2 → `isSyncInProgress() == true` → "Skipping iteration", returns without `commit()` → the batch (incl. `CUSTOMER/DELETED`) is dropped, position advances |
| 00:49:55,124 | sync completes — re-sends entities via fetchers (ADD-only), so it can't remove the now-stale customer |
| after | consumer keeps polling forward; `CUSTOMER/DELETED` is never reprocessed |

### The fix

Don't poll while the session isn't ready. If we never poll during the sync window, the Kafka position never advances past the event, so it stays in the queue and is delivered the moment the session is ready again — the same no-loss behaviour the Postgres edge gets for free.

`TbQueueConsumer` has no seek/pause, so rewinding the offset would be a high-blast-radius change. Instead this adds an opt-in poll gate to the shared `QueueConsumerManager` and wires only the edge consumer to it:

- `QueueConsumerManager` gets an optional `readinessCheck`; while it reports "not ready", the consumer loop skips polling (sleeps one poll interval and re-checks) instead of polling and then dropping.
- `KafkaEdgeGrpcSession` opts in via `isReadyToProcessGeneralEvents()` = `isConnected() && !isSyncInProgress() && !isHighPriorityProcessing`, and reuses that single predicate for the existing in-`processMsgs` guard too.

This is opt-in by design: every other consumer leaves `readinessCheck` unset, so it's always ready and its behaviour is byte-for-byte unchanged.

One caveat worth flagging for reviewers: if a sync ever ran longer than Kafka's `max.poll.interval.ms` (default 5 min), not polling would trigger a consumer-group rebalance; on rejoin the consumer resumes from the committed offset — still no data loss, just a rebalance. Edge syncs are normally seconds, so this is a rare edge case rather than a blocker.

### Validation

New `QueueConsumerManagerTest` in `common/queue`, covering:

- an event queued while the gate is closed is delivered once it opens — not dropped;
- the consumer isn't polled at all while not-ready, and resumes once ready;
- the default (no `readinessCheck`) path still polls and delivers immediately — the guarantee every other consumer relies on;
- the loop exits cleanly when interrupted mid-wait.

The two core tests are written so they fail on the old behaviour (without the gate it polls when it shouldn't, and the queued event is dropped) and pass with the fix. `common/queue` builds and all four tests are green.

No documentation changes needed — this is purely internal edge-consumer behaviour.

## General checklist

- [ ] You have reviewed the guidelines [document](https://docs.google.com/document/d/1wqcOafLx5hth8SAg4dqV_LV3un3m5WYR8RdTJ4MbbUM/edit?usp=sharing).
- [ ] [Labels](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/managing-labels#about-labels) that classify your pull request have been added.
- [ ] The [milestone](https://docs.github.com/en/issues/using-labels-and-milestones-to-track-work/about-milestones) is specified and corresponds to fix version.
- [ ] Description references specific [issue](https://github.com/thingsboard/thingsboard/issues).
- [x] Description contains human-readable scope of changes.
- [x] Description contains brief notes about what needs to be added to the documentation.
- [x] No merge conflicts, commented blocks of code, code formatting issues.
- [x] Changes are backward compatible or upgrade script is provided.
- [ ] Similar PR is opened for PE version to simplify merge. Crosslinks between PRs added. Required for internal contributors only.

## Front-End feature checklist

- [ ] Screenshots with affected component(s) are added. The best option is to provide 2 screens: before and after changes;
- [ ] If you change the widget or other API, ensure it is backward-compatible or upgrade script is present.
- [ ] Ensure new API is documented [here](https://github.com/thingsboard/thingsboard-ui-help)

## Back-End feature checklist

- [x] Added corresponding unit and/or integration test(s). Provide written explanation in the PR description if you have failed to add tests.
- [ ] If new dependency was added: the dependency tree is checked for conflicts.
- [ ] If new service was added: the service is marked with corresponding @TbCoreComponent, @TbRuleEngineComponent, @TbTransportComponent, etc.
- [ ] If new REST API was added: the RestClient.java was updated, issue for [Python REST client](https://github.com/thingsboard/thingsboard-python-rest-client) is created.
- [ ] If new yml property was added: make sure a description is added (above or near the property).

### CODE REVIEW

**Review by MazurenkoNick — COMMENTED** [review]
## Review summary

Reviewed 3 changed files in [Fixed edge event loss on Kafka edges during sync](https://github.com/thingsboard/thingsboard/pull/15756). Left 2 comment(s) inline.

The root-cause analysis in the description checks out. Verified against `TbKafkaConsumerTemplate`/`AbstractTbQueueConsumerTemplate`: `KafkaConsumer.poll()` advances the in-memory position regardless of commit, and `doCommit()` only does `commitSync()` (which matters on rebalance / a new session, not within the live consumer). So an uncommitted-but-polled batch is genuinely never re-read by that session — the "consumer position moves past it" claim is correct. The edge consumer is created with a `groupId`, so the `max.poll.interval.ms` rebalance caveat is also accurate.

The gate is a clean, opt-in design and is byte-for-byte unchanged for every other consumer (null `readinessCheck` ⇒ always ready). The comments below are about the one residual race the loop gate doesn't fully close, plus a minor API note — nothing blocking.

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java [hunk 1]
```diff
@@ -70,9 +70,9 @@ public KafkaEdgeGrpcSession(EdgeContextComponent ctx, TopicService topicService,
 
     private void processMsgs(List<TbProtoQueueMsg<ToEdgeEventNotificationMsg>> msgs, TbQueueConsumer<TbProtoQueueMsg<ToEdgeEventNotificationMsg>> consumer) {
         log.trace("[{}][{}] starting processing edge events", tenantId, edge.getId());
-        if (!isConnected() || isSyncInProgress() || isHighPriorityProcessing) {
+        if (!isReadyToProcessGeneralEvents()) {
```
Comment (MazurenkoNick) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java:75]: Now that the consumer loop gates on `isReadyToProcessGeneralEvents()` *before* polling, this guard becomes the second line of defense rather than the primary one. Worth keeping — but worth being explicit about what it now protects. The only way execution reaches here while not-ready is if readiness flips between the loop's gate check and this call, i.e. during the up-to-`pollInterval` `consumer.poll(...)`. When that happens, this returns without committing, but the batch has *already* been polled, so the position has advanced and those events are dropped — the same loss mode this PR fixes, just in a much narrower (~1s) window. Since `TbQueueConsumer` can't rewind, you can't fully close it here, but a one-line comment acknowledging the residual window (and that this duplicate condition is intentional defensive double-checking, not dead code) would stop a future reader from assuming the loop gate eliminated the race entirely.

File: common/queue/src/main/java/org/thingsboard/server/queue/common/consumer/QueueConsumerManager.java [hunk 2]
```diff
@@ -49,12 +51,13 @@ public class QueueConsumerManager<M extends TbQueueMsg> {
     @Builder
     public QueueConsumerManager(String name, MsgPackProcessor<M> msgPackProcessor,
                                 long pollInterval, Supplier<TbQueueConsumer<M>> consumerCreator,
-                                ExecutorService consumerExecutor, String threadPrefix) {
+                                ExecutorService consumerExecutor, String threadPrefix, BooleanSupplier readinessCheck) {
```
Comment (MazurenkoNick) [inline common/queue/src/main/java/org/thingsboard/server/queue/common/consumer/QueueConsumerManager.java:55]: The optional `readinessCheck` is the right shape — existing call sites pass nothing, `isReadyToProcess()` defaults to always-ready, so this stays backward compatible and is hard to misuse. Minor: a short Javadoc on the builder param spelling out the contract (e.g. "when supplied, the loop skips polling while this returns `false`, so the consumer position does not advance") would make the truthy-means-ready semantics obvious to the next consumer that wants to opt in, since a bare `BooleanSupplier` at a call site is otherwise opaque.

**Review by AndriiLandiak — COMMENTED** [review]
## Review summary

Reviewed 3 changed files in [Fixed edge event loss on Kafka edges during sync](https://github.com/thingsboard/thingsboard/pull/15756). Left 3 comment(s) inline.

The core change — an opt-in `readinessCheck` poll gate on `QueueConsumerManager`, wired only into the edge consumer — looks correct. The replacement predicate is boolean-equivalent to the code it supersedes (`!(a && !b && !c)` ⇔ `!a || b || c`), the gate is null-safe and leaves every other consumer byte-for-byte unchanged (verified: no direct constructor callers, all callers use the Lombok builder so the new param defaults to `null`/always-ready), and the documented consumer-rebalance trade-off when a sync exceeds `max.poll.interval.ms` is a reasonable, clearly-flagged edge case. The inline notes are about test coverage of the edge-specific wiring plus two minor design/maintainability points.

## Additional findings

This observation is about existing code outside the PR's diff — spotted while reading surrounding context.

- **application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java:147** — `processHighPriorityEvents()` sets `isHighPriorityProcessing = true` / `false` around `super.processHighPriorityEvents()` without a `try/finally`. Now that the new `readinessCheck` reads this flag to decide whether the consumer polls *at all*, a stuck `true` would park the consumer indefinitely rather than just skipping one in-line iteration as before. In practice the base method swallows every `Exception` internally, so today the flag can only stick on an `Error` (e.g. OOM) — but a `try/finally` would future-proof the toggle against base-method refactors now that its blast radius is larger.

---
*This review was auto-generated. Findings may contain errors — please verify before applying changes.*

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java [hunk 3]
```diff
@@ -126,6 +132,7 @@ public ListenableFuture<Boolean> processEdgeEvents() {
                         .consumerCreator(() -> tbCoreQueueFactory.createEdgeEventMsgConsumer(tenantId, edge.getId()))
                         .consumerExecutor(consumerExecutor)
                         .threadPrefix("edge-events-" + edge.getId())
+                        .readinessCheck(this::isReadyToProcessGeneralEvents)
```
Comment (AndriiLandiak) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java:135]: The new `QueueConsumerManagerTest` deterministically covers the generic gate, but the edge-specific wiring that actually fixes the bug — this `.readinessCheck(this::isReadyToProcessGeneralEvents)` line — has no deterministic test. If it were ever dropped, every unit test in this PR would still pass; the only thing that might catch the regression is the probabilistic `DashboardEdgeTest` BBT that surfaced the original ~50% flake. Worth a focused test (ideally an edge-level integration test) proving that an event queued while `isSyncInProgress()` survives until sync completes, so the actual fix is pinned rather than just the mechanism.

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java [hunk 4]
```diff
@@ -96,14 +98,18 @@ private void processMsgs(List<TbProtoQueueMsg<ToEdgeEventNotificationMsg>> msgs,
         }
     }
 
+    private boolean isReadyToProcessGeneralEvents() {
```
Comment (AndriiLandiak) [inline application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java:101]: Nice consolidation — this predicate is now the single source of truth for the poll gate (via `readinessCheck`), the `processMsgs` backstop, and `processEdgeEvents`. Since the gate and the backstop have to stay semantically identical for the "can't rewind" comment above to hold, a one-line note here that this method backs *both* the poll gate and the in-flight backstop would make that coupling discoverable from the definition rather than only from the call sites.

File: common/queue/src/main/java/org/thingsboard/server/queue/common/consumer/QueueConsumerManager.java [hunk 5]
```diff
@@ -102,6 +112,25 @@ private void consumerLoop(TbQueueConsumer<M> consumer) {
         }
     }
 
+    private boolean isReadyToProcess() {
+        return readinessCheck == null || readinessCheck.getAsBoolean();
```
Comment (AndriiLandiak) [inline common/queue/src/main/java/org/thingsboard/server/queue/common/consumer/QueueConsumerManager.java:116]: Minor: the nullable `readinessCheck` is handled with a null check here, which is what keeps all existing builder callers unchanged. An alternative is to default the field to an always-ready supplier (`() -> true`) in the builder, so this method could just call `readinessCheck.getAsBoolean()` and the nullable field disappears entirely. That trades a tiny constant allocation for one fewer null branch and a field that can't be left null — current form is fine, just flagging so the choice is deliberate.

Comment (AndriiLandiak) [reply common/queue/src/main/java/org/thingsboard/server/queue/common/consumer/QueueConsumerManager.java:116]: I prefer to leave like it's now, it's a pure-taste

File: application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java (hunk 4 above)
Comment (AndriiLandiak) [reply application/src/main/java/org/thingsboard/server/service/edge/rpc/KafkaEdgeGrpcSession.java:101]: No, it's already clear one
